Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/NetworkSearchController.java

```diff
@@ -131,53 +146,104 @@ public String getSearch(@ProjectIdentifier @PathVariable("projectUnixName") Stri
+            ITransformedNetwork transformedNetwork = transformationManager
+                    .getSearchTransformedNetwork(project.getProjectId(), conceptId, INetworkStatus.APPROVED);
+            if (transformedNetwork != null) {
+                publicSearchObject.setNodes(jsonCreator.getNodes(transformedNetwork.getNodes()));
+                publicSearchObject.setLinks(jsonCreator.getLinks(transformedNetwork.getLinks()));
+            }
+            if (transformedNetwork == null || transformedNetwork.getNodes().size() == 0) {
+                publicSearchObject.setNetworkEmpty(true);
+            }
+            return publicSearchObject;
+        };
+
+        Future<PublicSearchObject> future = executorService.submit(callable);
+        Random randomTokenGenerator = new Random();
+        Integer randomToken = randomTokenGenerator.nextInt(100);
+        while (searchResultMap.containsKey(randomToken)) {
+            randomToken = randomTokenGenerator.nextInt(100);
         }
-
-        model.addAttribute("jsonstring", json);
-        model.addAttribute("networkid", "\"\"");
-        model.addAttribute("project", project);
+        searchResultMap.put(randomToken, future);
+        model.addAttribute("token", randomToken);
+        model.addAttribute("unixName", projectUnixName);
         model.addAttribute("searchNodeLabel", searchNodeLabel);
         model.addAttribute("description", lemma);
-        model.addAttribute("unixName", projectUnixName);
+        model.addAttribute("project",project); 
 
         return "sites/networks/searchednetwork";
     }
+
+    /**
+     * This method checks if the request (corresponding to the tokenId) is
+     * processed. It returns a JSON containing the transformed network(if
+     * available) and sets the appropriate status: 1: Complete, 2:
+     * Invalid/Error, 3: In progress
+     * 
+     * @param projectUnixName
+     * @param tokenId
+     * @return JSON response with status and network details if availale
+     */
+
+    @ResponseBody
+    @RequestMapping(value = "sites/{projectUnixName}/networks/search/result", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
+    public PublicSearchObject getSearchTransformedNetwork(
+            @ProjectIdentifier @PathVariable("projectUnixName") String projectUnixName,
+            @RequestParam("tokenId") Integer tokenId) {
+
+        PublicSearchObject publicSearchObject = new PublicSearchObject();
+        if (!searchResultMap.containsKey(tokenId)) {
+            publicSearchObject.setStatus(2);
+            return publicSearchObject;
+        }
+
+        Future<PublicSearchObject> futureResult = searchResultMap.get(tokenId);
+        if (futureResult != null && futureResult.isDone()) {
+            try {
+                publicSearchObject = futureResult.get();
+                publicSearchObject.setStatus(1);
+                searchResultMap.remove(tokenId);
```

Comment:
also, in the service that manages pending searches I would use a cache that expires the searches at some point but will keep them around for a while (e.g. if the user reloads the page).